### ISSUE TITLE
gore/eval inserts "//line :xxx" lines into multiline raw string literals.

### ISSUE THREAD

**Issue by dmitshur** [sriram-srinivasan/gore#5]
``` bash
$ gore
Enter one or more lines and hit ctrl-D
fmt.Println(`This is a raw string literal,    
with newlines.`)
---------------------------------
This is a raw string literal,    
//line :2
with newlines.
```

The `func partition(code string) (topLevel string, nonTopLevel string)` function causes this.

**Comment by sriram-srinivasan** [comment]
Thanks for filing it as a bug; I'm aware of it. The code does not pay any attention to strings, nested strings, escaped characters, comments etc. If you would like to take a shot at it, I'd be very grateful!

Thanks.

**Comment by dmitshur** [comment]
I can see that fixing this isn't trivial. A quick solution to just look for starting and ending back quote symbols will break in heavier use cases. You don't want to parse Go code by hand.

Personally, I would fix it fast at the cost of temporarily taking out the "//line :xxx" comments altogether, because being able to Eval() code properly is a higher priority for me than to see correct line numbers in error messages.

However, I think a better solution might be to insert the "//line: xxx" comment only at the top of user code. This is a thought-in-process, but perhaps something like this might solve both problems.

Instead of, what you get currently:

``` go
package main
import "fmt"

//line :1
import "github.com/shurcooL/go-goon"
//line :2
import . "gist.github.com/5034040.git"

func main() {
     //line :3
goon.Dump(GetIndentedJson([]byte(`{
//line :4
  "id" : "user/17069122960193330967/source/com.google/post",
//line :5
  "title" : "\"post\" via Dmitri in Google Reader",
//line :6
  "author" : "Dmitri",
//line :7
  "updated" : 1367983706,
//line :8
  "direction" : "ltr",
//line :9
  "items" : [ ]
//line :10
}`)))

}

func __p(values ...interface{}){
    for _, v := range values {
             fmt.Printf("%+v\n", v)
    }
}
func __t(values ...interface{}){
    for _, v := range values {
             fmt.Printf("%T\n", v)
    }
}
```

What if the code were to be generated like this:

``` go
package main
import "fmt"

//line :1
import "github.com/shurcooL/go-goon"
import . "gist.github.com/5034040.git"

func main() {
    //line :3
     goon.Dump(GetIndentedJson([]byte(`{
  "id" : "user/17069122960193330967/source/com.google/post",
  "title" : "\"post\" via Dmitri in Google Reader",
  "author" : "Dmitri",
  "updated" : 1367983706,
  "direction" : "ltr",
  "items" : [ ]
}`)))

}

func __p(values ...interface{}){
    for _, v := range values {
             fmt.Printf("%+v\n", v)
    }
}
func __t(values ...interface{}){
    for _, v := range values {
             fmt.Printf("%T\n", v)
    }
}
```

**Comment by sriram-srinivasan** [comment]
> I can see that fixing this isn't trivial. A quick solution to just look for starting and ending back quote symbols will break in heavier use cases. You don't want to parse Go code by hand.

Well, it isn't really parsing Go code by hand, but just accounting for strings and comments, which requires a bit of a state machine. And quite possibly the same state machine that will work on Java, Go and C at the same time! 

However, I think a simpler option that'll work just fine for this particular case -- backticks and line numbers not working correctly together -- is this. Line numbers are only to provide feedback about errors in the original source, and if there were no errors to begin with, they need not be introduced at all. 
Recall (or notice!) that buildAndExec calls buildMain twice, the second time to recover from errors due to duplicate imported packages. If there are compiler errors other than duplicate imports, we introduce line numbers and rerun buildMain. Since this happens only in an error situation, you'll never see output with embedded line numbers.

**Comment by dmitshur** [comment]
> Well, it isn't really parsing Go code by hand, but just accounting for strings and comments, which requires a bit of a state machine. And quite possibly the same state machine that will work on Java, Go and C at the same time!

I think it's very unlikely that the required state machine will be simple, nor that it will work without bugs on all 3 languages.

> However, I think a simpler option that'll work just fine for this particular case -- backticks and line numbers not working correctly together -- is this. Line numbers are only to provide feedback about errors in the original source, and if there were no errors to begin with, they need not be introduced at all. Recall (or notice!) that buildAndExec calls buildMain twice, the second time to recover from errors due to duplicate imported packages. If there are compiler errors other than duplicate imports, we introduce line numbers and rerun buildMain. Since this happens only in an error situation, you'll never see output with embedded line numbers.

Now that sounds like an interesting idea that might work better.

I suppose the reason you prefer to include `//line: xxx` comments on every line is because the user might mix the top-level and main-func lines of code, so the linear sequence might be broken.

**Comment by sriram-srinivasan** [comment]
> I think it's very unlikely that the required state machine will be simple, nor that it will work without bugs on all 3 languages.

Hmm, not really. All one needs is a set of routines to look for the corresponding closing single/double/backtick quote, as well as for comments. Not much of a state machine.  Lack of proper string handling gives rise to all kinds of stupid bugs (like importing the ioutil package  when encountering the _string_ "ioutil.")

> I suppose the reason you prefer to include //line: xxx comments on every line is because the user might mix the top-level and main-func lines of code, so the linear sequences might be broken.

That's right. I wanted to be able to define types and structs and functions and imports in any order, and to interleave them with expressions, as befits an interactive experience. This means that lines have to be reorderable, and I need to remap the compiler error back to the original source line. Once strings are properly processed, line numbering is simple.

**Comment by dmitshur** [comment]
> Hmm, not really. All one needs is a set of routines to look for the corresponding closing single/double/backtick quote, as well as for comments. Not much of a state machine.

It's your call, feel free to give it a shot if you think it's a good approach.

I'd be wary of things like:

```
print(`\//*/`/*`*//*/\n\"\\"3"/'"\/*/)
```

Which is valid Go code that prints `\//*/`. Can you handle all the edge cases properly? What if Go syntax changes in a future version.

**Comment by sriram-srinivasan** [comment]
Fixed with commits 1c1e8efbf7f4d2f and the earlier 814089cf4
